test/HistoryTest.cs

```diff
@@ -118,136 +119,289 @@ public void SensitiveHistoryOptionalBehavior()
+                File.WriteAllText(newHistoryFilePath, string.Empty);
+                SetHistory(commandInputs);
+
+                historyItems = PSConsoleReadLine.GetHistoryItems();
+                Assert.Equal(commandInputs.Length, historyItems.Length);
+                for (int i = 0; i < commandInputs.Length; i++)
+                {
+                    Assert.Equal(commandInputs[i], historyItems[i].CommandLine);
+                }
+
+                // All commands are saved to the history file when 'ScrubSensitiveHistory' is set to 'false'.
+                text = File.ReadAllLines(newHistoryFilePath);
+                Assert.Equal(commandInputs.Length, text.Length);
+                for (int i = 0; i < text.Length; i++)
+                {
+                    Assert.Equal(commandInputs[i], text[i]);
+                }
+            }
+            finally
+            {
                 options.HistorySavePath = oldHistoryFilePath;
                 options.HistorySaveStyle = oldHistorySaveStyle;
                 options.AddToHistoryHandler = PSConsoleReadLineOptions.DefaultAddToHistoryHandler;
-                File.Delete(newHistoryFilePath);
-            }
+                File.Delete(newHistoryFilePath);
+            }
+        }
+
+        [SkippableFact]
+        public void SensitiveHistoryOptionalBehaviorWithScriptBlock()
+        {
+            TestSetup(KeyMode.Cmd);
+
+            // No history
+            SetHistory();
+            Test("", Keys(_.UpArrow, _.DownArrow));
+
+            var options = PSConsoleReadLine.GetOptions();
+            var oldHistoryFilePath = options.HistorySavePath;
+            var oldHistorySaveStyle = options.HistorySaveStyle;
+
+            // AddToHistoryHandler should be set to the default handler.
+            Assert.Same(PSConsoleReadLineOptions.DefaultAddToHistoryHandler, options.AddToHistoryHandler);
+
+            var newHistoryFilePath = Path.GetTempFileName();
+            var newHistorySaveStyle = HistorySaveStyle.SaveIncrementally;
+            Func<string, object> newAddToHistoryHandler_ReturnBool = LanguagePrimitives.ConvertTo<Func<string, object>>(
+                ScriptBlock.Create(@"
+                    param([string]$line)
+                    $line.Contains('gal')"));
+            Func<string, object> newAddToHistoryHandler_ReturnEnum = LanguagePrimitives.ConvertTo<Func<string, object>>(
+                ScriptBlock.Create(@"
+                    param([string]$line)
+                    if ($line.Contains('gal')) {
+                        [Microsoft.PowerShell.AddToHistoryOption]::MemoryOnly
+                    } elseif ($line.Contains('gmo')) {
+                        'SkipAdding'
+                    } else {
+                        [Microsoft.PowerShell.AddToHistoryOption]::MemoryAndFile
+                    }"));
+            Func<string, object> newAddToHistoryHandler_ReturnOther = LanguagePrimitives.ConvertTo<Func<string, object>>(
```

Comment:
I don't see an explicit test for `PSObject` return value. Maybe the result is always wrapped in `PSObject`, but it still seems wise to add an explicit test for that case on the off chance that the implementation changes to bypass the `PSObject` wrapper.

Reply:
Sure, will add an explicit test for `PSObject`.
I'm planning to change the implementation to not always wrap the result. It also seems to be a code reusing issue.